### PR TITLE
pre populate cache with views based on links with `turbo_preload` attribute

### PR DESCRIPTION
**Pull request by hey-leon**
First off, we ❤️  developing with hotwire and turbo at my workplace. So I want to thank all the contributors and the team.

Description
------------
At my workplace we have released a new mobile application using hotwire and coming from a react native application one part of the UX that felt a bit less refined was the first 0-300ms of navigation. When navigating to a page for a second time turbo loads a snapshot from cache which makes for a pretty snappy (pun intended) experience. However for hot paths we wanted to make this the reality on first visits.

Prior Art
---------
Before investigating this solution we did look at leveraging browser based prefetching. This can be achieved via the usage of `<link rel="prefetch" href="/app/path/here" />`. However our experience showed a couple ergonomic challenges using this for app screens:
* It does not appear to follow redirects in major browsers (expectations do not seem to be defined in specs).
* It relies on usage of the `Cache-control` header.
  * To leverage it you essentially need to completely hand over cache control to the browser which is not ideal for highly dynamic web apps.

The implementation in this PR makes it possible to pre populate the cache by marking links with `data-turbo-preload`. which makes for better user experience without needing to worry about stale views.

### CODE REVIEW

**Comment by seanpdoyle** [comment]
Thank you for opening this PR, the introduction of the `Preloader` class lots of potential!

I wonder, had you considered reading from an [`<a>` element's `[rel]` attribute](https://developer.mozilla.org/en-US/docs/Web/HTML/Element/a#attr-rel), instead of introducing a bespoke `[data-turbo-preload]` attribute?

There are existing semantics for values like [rel="prefetch"](https://developer.mozilla.org/en-US/docs/Web/HTML/Link_types/prefetch) and [rel="preload"](https://developer.mozilla.org/en-US/docs/Web/HTML/Link_types/preload), which could pose opportunities for progressive enhancement while gracefully degrading to built-in browser support when JS is disabled or Turbo fails to load.

File: src/core/drive/preloader.ts [hunk 1]
```diff
@@ -0,0 +1,48 @@
...
+
+  preloadOnLoadLinksForView(element: Element) {
+    const links = element.querySelectorAll('a[data-turbo-preload="load"]')
+
+    for (const link of links) {
+      this.preloadURL(link)
+    }
+  }
+
+  preloadURL(link: Element) {
```
Comment (seanpdoyle) [inline src/core/drive/preloader.ts:35]: If this method accepted [HTMLAnchorElement](https://developer.mozilla.org/en-US/docs/Web/API/HTMLAnchorElement) as a type, we could read the [link.href](https://developer.mozilla.org/en-US/docs/Web/API/HTMLAnchorElement/href) directly, without the need to call `expandURL` or `|| ""`.

File: src/core/drive/preloader.ts [hunk 2]
```diff
@@ -0,0 +1,48 @@
...
+
+    if (this.snapshotCache.has(url)) {
+      return
+    }
+
+    fetch(url.toString(), { headers: { 'VND.PREFETCH': 'true', 'Accept': 'text/html' } })
+      .then(res => res.text())
+      .then(responseText => PageSnapshot.fromHTMLString(responseText))
+      .then(snapshot => this.snapshotCache.put(url, snapshot))
+      .catch(() => {})
```
Comment (seanpdoyle) [inline src/core/drive/preloader.ts:46]: This codebase utilizes [`async`/`await` syntax](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Statements/async_function), which enables this to be implemented without chaining:

```js
const response = await fetch(url.toString(), { headers: { 'VND.PREFETCH': 'true', 'Accept': 'text/html' } })
const responseText = await response.text()
const snapshot = PageSnapshot.fromHTMLString(responseText)

this.snapshotCache.put(url, snapshot)
```

File: src/core/drive/preloader.ts (hunk 1 above)
Comment (hey-leon) [reply src/core/drive/preloader.ts:35]: Thanks for the feedback @seanpdoyle I'll have a go at refining the types.

File: src/core/drive/preloader.ts (hunk 2 above)
Comment (hey-leon) [reply src/core/drive/preloader.ts:46]: Yes that does seem to be the convention. I will tidy this up 👍. Thank you.

**Comment by hey-leon** [comment]
> Thank you for opening this PR, the introduction of the `Preloader` class has a lot of potential!
> 
> I wonder, had you considered reading from an [`<a>` element's `[rel]` attribute](https://developer.mozilla.org/en-US/docs/Web/HTML/Element/a#attr-rel), instead of introducing a bespoke `[data-turbo-preload]` attribute?
> 
> There are existing semantics for values like [rel="prefetch"](https://developer.mozilla.org/en-US/docs/Web/HTML/Link_types/prefetch) and [rel="preload"](https://developer.mozilla.org/en-US/docs/Web/HTML/Link_types/preload), which could pose opportunities for progressive enhancement while gracefully degrading to built-in browser support when JS is disabled or Turbo fails to load.

Good point 

> 

This is a great thought. The "prefetch" and "preload" types are not currently supported by anchor tags (link only) at this point. See [this](https://developer.mozilla.org/en-US/docs/Web/HTML/Link_types) table for rel type support. But based on moz docs it appears it is at least being considered [here](https://developer.mozilla.org/en-US/docs/Web/HTTP/Link_prefetching_FAQ#are_anchor_a_tags_prefetched).

What is your thoughts on adopting this pre-emptively. I am open to the refactor if we think this is a good approach?

File: src/core/drive/preloader.ts (hunk 2 above)
Comment (hey-leon) [reply src/core/drive/preloader.ts]: cfd830d0df4c46a4383e620117005e113dd77472

File: src/core/drive/preloader.ts (hunk 1 above)
Comment (hey-leon) [reply src/core/drive/preloader.ts]: https://github.com/hotwired/turbo/pull/552/commits/8e353855278743a3585fffc13b2bf0d27fbd51eb

File: src/core/drive/preloader.ts [hunk 3]
```diff
@@ -28,23 +27,25 @@ export class Preloader {
     const links = element.querySelectorAll('a[data-turbo-preload="load"]')
```
Comment (seanpdoyle) [inline src/core/drive/preloader.ts:27]: Does executing this query with additional type information help avoid the `instances` check?

```js
for (const link of element.querySelectorAll<HTMLAnchorElement>('a[data-turbo-preload="load"]')) {
  this.preloadURL(link)
}
```

Comment (hey-leon) [reply src/core/drive/preloader.ts]: Resolved in https://github.com/hotwired/turbo/pull/552/commits/15fedf0af59c68e77325c5c60bf45949437d73ac

**Comment by rik** [comment]
I'm concerned about Turbo re-purposing existing HTML attributes. Currently, `rel=preload` and `rel=prefetch` aren't allowed on `<a>` by [the HTML spec](https://html.spec.whatwg.org/multipage/links.html#linkTypes:link-type-prefetch-2). If the HTML spec wants to change to allow it, Turbo might get in the way.

On top of that, I'm not sure this feature is needed in Turbo. You can add `<link rel=prefetch>` in your HTML (or in JS) and supporting browsers will prefetch them. So the fetch call from Turbo when visiting should be instant. Maybe just a polyfill for Safari [until they implement prefetching](https://bugs.webkit.org/show_bug.cgi?id=194539)?

**Comment by hey-leon** [comment]
> I'm concerned about Turbo re-purposing existing HTML attributes. Currently, `rel=preload` and `rel=prefetch` aren't allowed on `<a>` by [the HTML spec](https://html.spec.whatwg.org/multipage/links.html#linkTypes:link-type-prefetch-2). If the HTML spec wants to change to allow it, Turbo might get in the way.

@rik you are correct that they are not supported on anchors at this point. However mozilla's [documentation](https://developer.mozilla.org/en-US/docs/Web/HTTP/Link_prefetching_FAQ#are_anchor_a_tags_prefetched) demonstrates an openness to to accept it. Aside from this, the more I get in the weeds on this PR I am realizing there are some fundamental differences between the changes in this PR and `link[rel="prefetch"]` and I think that is probably a good enough reason to change to a custom attributes.

> On top of that, I'm not sure this feature is needed in Turbo. You can add `<link rel=prefetch>` in your HTML (or in JS) and supporting browsers will prefetch them. So the fetch call from Turbo when visiting should be instant. Maybe just a polyfill for Safari [until they implement prefetching](https://bugs.webkit.org/show_bug.cgi?id=194539)?

After my exploration I am not sure if `link[rel="prefetch"]` is an awesome time for the use case outlined in this PR. Leveraging `link[rel="prefetch"]` was the first approach I looked at before authoring this PR. I too would prefer to leverage the native browser functionality. However the ergonomics seem to be quite challenging. There are two main problems:
- It doesn't appear to follow redirects (however this seems to be undefined in spec)
- It sensibly but painfully abides the cache-control header.

So personally I feel there is a benefit with being able to preload a preview while not having to hand cache control to the browser.

File: src/core/drive/preloader.ts [hunk 4]
```diff
@@ -0,0 +1,54 @@
...
+import { PageSnapshot } from "./page_snapshot"
+import { SnapshotCache } from "./snapshot_cache"
+
+export interface PreloaderDelegate {
+  readonly navigator: Navigator
+}
+
+export class Preloader {
+  readonly delegate: PreloaderDelegate
+  readonly selector: string = 'a[data-turbo-preload="true"]'
```
Comment (ghiculescu) [inline src/core/drive/preloader.ts:11]:
```suggestion
  readonly selector: string = 'a[data-turbo-preload]'
```

I think this is the preferred style here when a truthy attribute is expected, for example see https://github.com/hotwired/turbo/blob/8221c1e5f2021c41c3ecada0dde55bae9ee18380/src/core/snapshot.ts#L29

**Comment by hey-leon** [comment]
Hey turbo team, I wanted to circle back here as I have not stayed on top of this.

I can say that we have been using this in production to preload our hottest app paths with success. I am wondering if there is interest in merging this feature or a revision of this?

@seanpdoyle do you have any opinions?

**Comment by ghiculescu** [comment]
@dhh do you have any feedback on the concept here?

**Comment by silentjay** [comment]
Hoping this gets merged. Some of us have already implemented pre-fetching via 3rd party scripts mentioned in this old issue https://github.com/hotwired/turbo/issues/174 would be great to have it built in.

**Comment by dhh** [comment]
This is great. Agree it would be nice to lean on a href's support when that's there, but don't think we should get ahead of the spec. If it ends up happening, it wouldn't be a big deal to go from data-turbo-preload to whatever that is. But as with the link preloads, it might well be that the cache-control dynamics don't match with what we're trying to do here.

**Comment by ghiculescu** [comment]
Should we update https://github.com/hotwired/turbo-site to note this feature? Maybe as a new item in https://turbo.hotwired.dev/handbook/drive ?

**Comment by hey-leon** [comment]
> This is great. Agree it would be nice to lean on a href's support when that's there, but don't think we should get ahead of the spec. If it ends up happening, it wouldn't be a big deal to go from data-turbo-preload to whatever that is. But as with the link preloads, it might well be that the cache-control dynamics don't match with what we're trying to do here.

Thanks for taking the time to understand the PR @dhh. I think your analysis pretty spot on as well.

> Should we update https://github.com/hotwired/turbo-site to note this feature? Maybe as a new item in https://turbo.hotwired.dev/handbook/drive ?
> Yes please do add a PR we can ship with 7.2.0 👍

Good suggestion. I will put something together 👍

**Comment by clementmas** [comment]
Are the linked pages all preloaded on page load or only when each link becomes visible?

If not, I think that could be a great addition in a new PR. Using Intersection Observers. I think Next.js and Nuxt.js do that automatically.

**Comment by hey-leon** [comment]
> Are the linked pages all preloaded on page load or only when each link becomes visible?
> 
> If not, I think that could be a great addition in a new PR. Using Intersection Observers. I think Next.js and Nuxt.js do that automatically.

That could be a pretty cool idea @clementmas

**Comment by pfeiffer** [comment]
> > Are the linked pages all preloaded on page load or only when each link becomes visible?
> > If not, I think that could be a great addition in a new PR. Using Intersection Observers. I think Next.js and Nuxt.js do that automatically.
> 
> That could be a pretty cool idea @clementmas

In that case I'd prefer to have links annotated with `data-turbo-preload="lazy"` to opt-in to the lazy-preloading behavior. I believe that links explicitly hinted with `data-turbo-preload="true"` should still be preloaded, even outside of the viewport, as those links could be visited programatically.

A different approach to this could be to implement the functionality in this PR using MutationObserver instead of preloading links after a frame visit or on document load. This would allow developers to define their own behavior on top, eg. by implementing an Interaction Observer that sets the `data-turbo-preload` as links scroll into view.

It could also be used to add `data-turbo-preload` when the user has a "hover intent" (ie. like https://www.npmjs.com/package/hoverintent).

**Comment by hey-leon** [comment]
> > > Are the linked pages all preloaded on page load or only when each link becomes visible?
> > > If not, I think that could be a great addition in a new PR. Using Intersection Observers. I think Next.js and Nuxt.js do that automatically.
> > 
> > 
> > That could be a pretty cool idea @clementmas
> 
> In that case I'd prefer to have links annotated with `data-turbo-preload="lazy"` to opt-in to the lazy-preloading behavior. I believe that links explicitly hinted with `data-turbo-preload="true"` should still be preloaded, even outside of the viewport, as those links could be visited programatically.
> 
> A different approach to this could be to implement the functionality in this PR using MutationObserver instead of preloading links after a frame visit or on document load. This would allow developers to define their own behavior on top, eg. by implementing an Interaction Observer that sets the `data-turbo-preload` as links scroll into view.
> 
> It could also be used to add `data-turbo-preload` when the user has a "hover intent" (ie. like https://www.npmjs.com/package/hoverintent).

Yes I don't think there is any reason to remove the current functionality for our purpose I'm not sure we would need the intersection observer but I cannot see why we can't have both approaches. Having said that if we wanted to keep the API simple I don't see any downside to the intersection observer approach for us. E.g. most things we are preloading are defined in fixed navbars and they usually load in such a small time frame the time between pressing the link and it coming into frame it would likely be loaded.

**Pull request merged**